File: graph_analysis/make_sedges.rb
require 'time'
require 'db.rb'
# running in rails environment
require 'app/models/cmodule.rb'
require 'app/models/sedge.rb'

_start = Time.now.to_f
relations = Hash.new
Cmodule.find(:all).each do |mod|
  relations[mod.name] ||= Array.new
  relations[mod.name] << mod.site_id
end

#
##
# I am sure I can do this faster
#


GC.start

_similar = Hash.new

# now add a ton of edges:
relations.keys.each do |k|
  p k,relations[k].size
  c = 0
  relations[k].each do |item|
    relations[k].each do |in_item|
      next if (in_item == item)
      
      _similar["#{item}.#{in_item}"] ||= 0 
      _similar["#{item}.#{in_item}"] += 1 
      #print "."
      c +=1
    end
    print "#{relations[k].size * relations[k].size - c}\n"
  end
end

_similar.keys.each do |k|
      p k, _similar[k]
      (item, in_item) = k.split(".")
      me = Sedge.new
      me.in_site_id = item.to_i
      me.out_site_id = in_item.to_i
      me.common = _similar[k]
      me.save
end 

print "Finished in: #{Time.now.to_f - _start}\n"

File: misc_graphing_scripts/make_sedges.py
import MySQLdb
import time

dbh = MySQLdb.connect(host='localhost', db='php_info', user='root' )

c = dbh.cursor()


_start = time.time()
relations = {}

res = c.execute("select max(id) from sites")
(id,) = c.fetchone()
print id 

#build big table:
_big_list = array()
for c in xrange(0, id+1):
  _big_list.append(array())
  for j in xrange(0, id+1):
    _big_list[c].append(0)

res = c.execute("select name, site_id from cmodules")
for (name,site_id) in c:
  if(not relations.get(name,False)):
    relations[name] = []
  relations[name].append(site_id)

c.close()

_similar = {}

# now add a ton of edges:
for k,v in relations.items():
  print k, len(v)
  c = 0
  for i in v:
    for j in v:
      if(i != j):
        key = "%s.%s" %(i,j)
        if(not _similar.get(key, None)):
          _similar[key] = 0
        _similar[key] +=1
      c +=1
    if( c % int(len(v)*len(v) /4) == 0 ):
      print str( len(v) * len(v) - c) 

c = dhb.cursor()
for k,v in _similar:
      print k, v
      (item, in_item) = k.split(".")
      c.execute("insert into sedges (in_site_id, out_site_id, common) values(%s, %s, %i)", (item, in_item, v))

print "Finished in: %s" %( time.time() -  _start)
